### PR TITLE
Finalize fixes for semantic model and IDE support for merging

### PR DESCRIPTION
**Pull request by agocke**
This PR should clean up the outstanding required compiler tests for merging, as well as IDE support. The required part of the IDE checklist is as follows:

image

### CODE REVIEW

**Comment by VSadov** [comment]
Why picture?
I think in the past we were just pasting a copy of tables form https://github.com/dotnet/roslyn/wiki/Manual-Testing

It does result in a lot of N/A boxes, but it is easy to see that nothing was missed.

**Comment by agocke** [comment]
@VSadov I'm not sure if the Github Markdown parser changed, but copying that code doesn't work anymore. It doesn't result in a formatted table. Fixing up all the errors by hand seemed like way too much work for what amounts to a check box.

**Comment by agocke** [comment]
@dotnet/roslyn-ide Please review when you get a chance.

**Comment by jcouv** [comment]
I wonder if this fixes https://github.com/dotnet/roslyn/issues/25063 too

**Comment by agocke** [comment]
@jcouv I don't think so. The code I added only runs in a `foreach`. It looks like it's still silently crashing for ref locals.

**Comment by agocke** [comment]
ping @dotnet/roslyn-ide for review. This feature needs to be merged into 15.7.x to make preview 2.

File: src/Compilers/CSharp/Test/Symbol/Compilation/SemanticModelAPITests.cs [hunk 1]
```diff
@@ -16,6 +16,87 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests
...
+            {
+                return;
+            }
+        }
+    }
+}");
+            comp.VerifyDiagnostics();
+            var tree = comp.SyntaxTrees.Single();
+            var root = tree.GetCompilationUnitRoot();
+            var rx = root.DescendantNodes().OfType<IdentifierNameSyntax>().Last();
```
Comment (jasonmalinowski) [inline src/Compilers/CSharp/Test/Symbol/Compilation/SemanticModelAPITests.cs]: Do we ever also call GetDeclaredSymbol() on the first one and verify symbols are the same?

File: src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs [hunk 2]
```diff
@@ -4612,6 +4612,33 @@ void M()
...
+            // introduce local still introduces a local, not a ref-local.
+            await TestInRegularAndScriptAsync(@"
+class C
+{
+    void M(int x)
+    {
+        ref int y = ref x;
+        M2([|(y = ref x)|]);
+    }
+    void M2(ref int y) { }
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs]: As discussed in person, this case doesn't compile anyways and needs a "ref" at the call site.

**Comment by agocke** [comment]
@jasonmalinowski Fixed. I noticed that the IDE doesn't currently seem to support extracting expressions being passed/assigned by-ref. For example, extract local does not work for this case:

```C#
 M(ref [|M2()|]);

static void M(ref int x) { }
static ref int M2() => throw new Exception();
```

**Comment by jasonmalinowski** [comment]
@agocke I'm not surprised. Not sure if it ever worked, possibly? It's such a niche scenario I could imagine nobody until you has noticed.

File: src/Compilers/CSharp/Test/Symbol/Compilation/SemanticModelAPITests.cs [hunk 3]
```diff
@@ -65,6 +65,8 @@ void M(int x)
             var rxSymbol = Assert.IsAssignableFrom<ILocalSymbol>(rxInfo.Symbol);
             Assert.True(rxSymbol.IsRef);
             Assert.Equal(RefKind.RefReadOnly, rxSymbol.RefKind);
+            var rxDecl = root.DescendantNodes().OfType<ForStatementSyntax>().Single().Declaration;
+            Assert.Equal(model.GetDeclaredSymbol(rxDecl.Variables.Single()), rxSymbol);
```
Comment (jasonmalinowski) [inline src/Compilers/CSharp/Test/Symbol/Compilation/SemanticModelAPITests.cs]: Consider: Assert.Same.

File: src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs [hunk 4]
```diff
@@ -4628,17 +4628,33 @@ void M(int x)
...
     }
-    void M2(ref int y) { }
+    void M2(int p) { }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceVariable)]
+        public async Task TestIntroduceLocalInRefCallRefExpression()
+        {
+            // Cannot extract expressions passed by ref
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharpTest/CodeActions/IntroduceVariable/IntroduceVariableTests.cs]: Might want to comment that this is more a bug than anything else, but this at least correctly asserts we don't crash and die.

**Pull request merged**